pkg/proxy/iptables/proxier.go

```diff
@@ -656,6 +664,36 @@ func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
 	proxier.deleteEndpointConnections(staleConnections)
 }
 
+// needToUpdate returns if the input endpoints is different from the last batch
+// This is calculated by comparing the checksum of EndpointSubsets.
+func (proxier *Proxier) needToUpdate(allEndpoints []api.Endpoints) bool {
+	checksumMap := make(map[string]uint32)
+	hash := fnv.New32a()
+
+	for _, ep := range allEndpoints {
+		hashutil.DeepHashObject(hash, ep.Subsets)
+		checksumMap[fmt.Sprintf("%s/%s", ep.Namespace, ep.Name)] = hash.Sum32()
+	}
+
+	needToUpdate := false
+	if len(checksumMap) == len(proxier.endpointChecksums) {
+		for ep, checksum := range checksumMap {
+			if checksum != proxier.endpointChecksums[ep] {
+				needToUpdate = true
+				break
+			}
+		}
+	} else {
+		needToUpdate = true
+	}
+
+	if needToUpdate {
+		proxier.endpointChecksums = checksumMap
```

Comment:
I guess this method is not thread safe. If you call it from both OnServiceUpdate and OnEndpointUpdate, can they race? or are they always invoked serially? If there's a possibility of a race you should wrap this in setEndpointChecksums, getEndpointChecksums which does the expected set/get under a lock